clients/roscpp/src/libros/connection.cpp

```diff
@@ -337,7 +339,10 @@ void Connection::drop(DropReason reason)
 
   if (did_drop)
   {
-    drop_signal_(shared_from_this(), reason);
```

Comment:
Should this line better be moved up into the `if (!dropped_)` block to avoid the mutex being locked twice?

[I would assume there was a reason why the call was originally not within the locked mutex scope and am a bit concerned this could lead to a regression.]

Reply:
If there is no PR https://github.com/ros/ros_comm/pull/1950, could use only one lock and move the drop_signal_ and close() up.
Now #1950 has been merged. So it seems that it needs lock twice here, the modification will be as follows,

```
void Connection::drop(DropReason reason)

  if (did_drop)
  {
    transport_->close();
    {
      boost::recursive_mutex::scoped_lock lock(drop_mutex_);
      drop_signal_(shared_from_this(), reason);
    }
  }

```

moving the transport->close() before the first lock might be work, but I am not sure if there is a restriction.

Reply:
> [I would assume there was a reason why the call was originally not within the locked mutex scope and am a bit concerned this could lead to a regression.]

Think so too. Not very clear about the original thought. So keep the original logic here but to add lock if necessary.